File: python/Array/SubarraySumEqualsK.py
class Solution:
    def subarraySum(self, nums, k):
        # count=0
        # for i in range(0,len(nums)):
        #     sum=0
        #     for j in range(i,len(nums)):
        #         sum+=nums[j]
        #         if sum==k:
        #             count+=1
        # return count
        mp={}
        sum=0
        count=0
        mp[0]=1
        for i in range(0,len(nums)):
            sum+=nums[i]
            if sum-k in mp:
                count+=mp[sum-k]
            if sum in mp:
                mp[sum]+=1
            else:
                mp[sum]=1
        return count

n=int(input())    
nums=list(map(int,input().split()))
k=int(input())
obj=Solution()
result=obj.subarraySum(nums,k)
print(result)

File: cpp/Array/SubarraySumEqualsK.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution{
    public:
        int subarraySum(vector<int>& nums, int n, int k)
        {
            // int count=0;
            // for(int i=0;i<n;i++)
            // {
            //     int sum=0;
            //     for(int j=i;j<n;j++)
            //     {
            //         // for(int k=i;k<=j;k++)
            //         // {
            //         //      .....
            //         // }
            //         sum+=nums[j];
            //         if(sum==k)
            //         {
            //             count++;
            //         }
            //     }
            // }
            // return count;

            map<int,int>mp;
            int sum=0;
            int count=0;
            for(int i=0;i<n;i++)
            {
                sum+=nums[i];
                if(sum==k)
                {
                    count++;
                }
                if(mp.find(sum-k)!=mp.end())
                {
                    count+=mp[sum-k];
                }
                mp[sum]++;
            }
            return count;
            
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int n;
    cin>>n;
    int k;
    cin>>k;
    vector<int>a(n);
    for(int i=0;i<n;i++)
    {
        cin>>a[i];
    }
    Solution obj;
    cout<<obj.subarraySum(a,n,k);
}